c2corg_api/tweens/rate_limiting.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import datetime
+import pytz
+
+from pyramid.httpexceptions import HTTPBadRequest, HTTPTooManyRequests
+from c2corg_api.emails.email_service import get_email_service
+from c2corg_api.views import http_error_handler
+from c2corg_api.models import DBSession
+from c2corg_api.models.user import User
+
+log = logging.getLogger(__name__)
+
+
+def rate_limiting_tween_factory(handler, registry):
+    """ Add a rate limiting protection on write requests.
+    """
+
+    def tween(request):
+
+        log.debug('RATE LIMITING FOR METHOD ' + request.method)
+
+        # Only write requests are considered for rate limiting.
+        if request.method not in ['POST', 'PUT', 'DELETE']:
+            return handler(request)
+
+        if request.authorization is None:
+            # See comment of similar block in jwt_database_validation tween
+            return handler(request)
+
+        user = DBSession.query(User).get(request.authenticated_userid)
+        if user is None:
+            return http_error_handler(
+                HTTPBadRequest('Unknown user'), request)
+
+        now = datetime.datetime.now(pytz.utc)
+        if user.ratelimit_reset is None or user.ratelimit_reset < now:
+            # No window exists or it is expired: create a new one.
+            span = int(registry.settings.get('rate_limiting.window_span'))
+            limit = int(registry.settings.get(
+                'rate_limiting.limit_robot' if user.robot else
+                'rate_limiting.limit_moderator' if user.moderator else
+                'rate_limiting.limit'))
+            user.ratelimit_reset = now + datetime.timedelta(seconds=span)
+            user.ratelimit_limit = limit
+            user.ratelimit_remaining = limit - 1
+        elif user.ratelimit_remaining:
+            user.ratelimit_remaining -= 1
+        else:
+            # User is rate limited
+
+            # Count how many windows the user has been rate limited
+            # and block them is too many.
+            current_window = user.ratelimit_reset
+            if user.ratelimit_last_blocked_window != current_window:
+                user.ratelimit_last_blocked_window = current_window
+                user.ratelimit_times += 1
+
+                max_times = int(
+                    registry.settings.get('rate_limiting.max_times'))
+                if user.ratelimit_times > max_times:
+                    user.blocked = True
+
+                # An alert message is sent to the moderators
```

Comment:
Please note sending the email postpones a little the "429 Too Many Requests" response.
If it is too long, a timeout is hit and I think it causes a 500 Internal Server Error (which is quite the same than the 429 error :P for the end user).

I don't know if it is possible to send the email asynchronously. Or maybe I could embed the 2 lines below in a try-except block?

Reply:
Seems it is possible to use another process to send the mail (through a queue): https://docs.pylonsproject.org/projects/pyramid_mailer/en/latest/#queue

Reply:
Thanks for the tip!

To be honest I feel a bit lazy doing this because it adds quite some work to set the separate process whereas sending such mails should not be that frequent, especially when the result is refusing a document edition anyway :)